src/array.jl

```diff
@@ -943,16 +943,17 @@ function Base.sort!(v::CategoricalVector;
 
     # compute the order in which to read from counts
     ord = Base.Sort.ord(lt, by, rev, order)
-    index = eltype(v) >: Missing ? [missing; v.pool.valindex] : v.pool.valindex
     seen = counts .> 0
     anymissing = eltype(v) >: Missing && seen[1]
-    perm = sortperm(view(index, seen), order=ord)
-    nzcounts = counts[seen]
+    levs = eltype(v) >: Missing ? [missing; v.pool.valindex] : v.pool.valindex
+    sortedlevs = sort(Vector(view(levs, seen)), order=ord)
```

Comment:
why not use? `sort!(levs[seen], order=ord)`

Reply:
We need a `Vector` to avoid calling ourselves and getting a stack overflow. But yes `sort!` should work.

Reply:
`levs[seen]` would create a new vector anyway - right? But this is minor I guess so it can stay as is.

Reply:
Actually there's a weird behavior which creates a `CategoricalVector` with logical indexing. Need to investigate that...

Reply:
This is so tricky that I ended up filing a design issue against Base: https://github.com/JuliaLang/julia/issues/36106.